### ISSUE TITLE
npm@5 outdated -g shows nothing

### ISSUE THREAD

**Issue by Ashot-KR** [npm/npm#16840]
#### What's going wrong?
**npm@4:** `npm outdated -g` output:
```
Package          Current  Wanted         Latest  Location
gemini     5.0.0-alpha.1  4.19.3  5.0.0-alpha.1
npm                4.6.1   5.0.0          5.0.0
uglify-js          3.0.3   3.0.3         3.0.14
``` 
  
**npm@5:** `npm outdated -g` outputs nothing. Operation takes about 10-20 seconds and silently
 exits without any output
### supporting information:

 - `npm -v` prints: 5.0.0
 - `node -v` prints: 8.0.0
 - `npm config get registry` prints: https://registry.npmjs.org/
 - macOS 10.12.4

**Comment by moimael** [comment]
I also have the same issue after upgrading.

**Comment by ingro** [comment]
Same thing for me on Windows 10.

**Comment by skiabox** [comment]
Same problem here with npm 5.0.1
It is probably a bug.

**Comment by themaxmarchuk** [comment]
Same issue for me (on Windows 10) I usually have the log level set to http and when using npm 5.0.0 and 5.0.1 no http requests are logged.

**Comment by tiru2wtf** [comment]
Same here on Mac OS 10.12.5, node v8.0.0, npm v5.0.1

**Comment by sfabriece** [comment]
same here windows 10, npm v5.0.1

**Comment by fsievers** [comment]
Maybe related to this:
npm update -g
does not update modules. This may be a consequence of the face, that `npm outdated -g` does not find outdated packages.

Tested with npm 5.0.0 and 5.0.1
Node 6.10.2
Linux (Ubuntu 14.04)

**Comment by skiabox** [comment]
For the time being and till the broken npm is fixed I am using npm-check -g command.

**Comment by mcibique** [comment]
Same issue for me in npm 5.0.3 on Ubuntu and Win10. I quickly debugged the code and I can see that npm5 is able to load the globally installed packages and creates the tree, but then it marks all packages as extraneous and skips checking all of them (it happens on this [line in code](https://github.com/npm/npm/blob/v5.0.3/lib/outdated.js#L207)). 

I have checked the same code on npm@4 and none of global packages are marked extraneous on the same line, so they can be later checked for newer version against registry. 

I see that the logic inside `isExtraneous` function has changed in npm@5 13 days ago in this commit: https://github.com/npm/npm/commit/3116cfecdaaf3b76abe44ae1e3e08cce8bf8cfe1#diff-2986d073339d9e2bb0622129b5f9f93aL27. The removal of line 27 causes this issue.

**Comment by dagilleland** [comment]
**Work-Around:** While `npm outdated -g` isn't working, you can still

1. **`npm ls -g --depth=0`** list global packages,
1. **`npm i -g <package-name>`** just run an install of the package you're concerned about, and
1. **`npm ls -g --depth=0`** list global packages again to see if they have changed.

It won't tell you *which* ones are out-of-date, but you can at least see what you have globally and update it if you know/suspect there are new versions. (Helped me as a relative newbie)

**Comment by darkmorpher** [comment]
https://github.com/npm/npm/issues/17209 may be related

**Comment by alexshin** [comment]
npm@5.0.3 on MacOS 10.12.5 outputs nothing too =(

**Comment by thewashingtonhua** [comment]
Same here with npm@5.0.3 on macOS 10.12.5

`npm outdated` shows nothing when there should be.
`npm install` failed.
`npm uninstall` incomplete leaving a link causing 'too many levels of symbolic links'. which made npm unusable when I tried to upgrade to 5.0.4. I have to download the source code and manually bring it back to 4.6.1.

So basicly current version (5.0.x) of npm is totally not functional. Had to downgrade. PLEEEEEEEASE fix ASAP. The world depends on you guys.

**Comment by arnojernej** [comment]
Same here with npm@5.0.4 node@8.1.2 macOS 10.12.5

**Comment by ankur-p** [comment]
Same issue: npm@5.0.4, node@6.11.0, macOS 10.12.5.  I had downgraded to 4.6.1 from 5.0.3 which was having the same problem.  4.6.1 worked fine with no other changes to my environment.  Looks like 5.0.4 didn't fix it, so I'll be reverting back to 4.6.1.

**Comment by motss** [comment]
I've come across a great package [npx](https://github.com/zkat/npx) and have been using it since then. The great things about it is that I no longer need to install so many global dependencies except for those dependencies do not name exactly the same as the commands to run them such as `fkill-cli`, I will install it globally.

This little script here might help: https://gist.github.com/motss/85d90d3390bb0b13e9ec9be61c71af79

___P/S: I'm no expert in writing shell script and feel free to improve it if you want to! 😄___

**Comment by cs-miller** [comment]
same problem, npm@5.0.4 node@8.1.2 sierra@10.12.5

**Comment by dukegod** [comment]
same problem

**Comment by fsievers** [comment]
~~Seems to be fixed in 5.0.4-canary.13~~

Edit: Hmm, sry. Seems like I run npm4 when I was thinking I run npmc@5.0.4-canary.13.

**Comment by riyadhalnur** [comment]
Issue still persists on 5.1.0

**Comment by riyadhalnur** [comment]
Issue resolved in 5.2.0

**Comment by motss** [comment]
This is confirmed. Just tested it immediately got the update and it worked flawlessly (finally) 😭

**Comment by dukegod** [comment]
@riyadhalnur  it resolved!

**Comment by ankur-p** [comment]
Resolved in my environment as well.  Thanks!

**Comment by motss** [comment]
Any Windows 10 user can verify if the fix has solved your issue? It did solve on my macOS machine but not Windows 10.

**Comment by alaindeurveilher** [comment]
I have npm 5.3.0 (with node 8.2.1) and the issue is not fixed yet: `npm outdated` still does not output anything at all.

I was trying a workaround:
```bash
nvm use 7.10.0
npm outdated
```
but the command outputs the global packages instead of my local packages!

So I guess I will make this other and painful workaround instead:
```
nvm use 8.2.1
cd my-project
rm -rf node_module
yarn
yarn outdated
```
then check the new packages release notes, and optionally update them if I judge necessary.

When I am OK with my updated package.json:
```
rm -rf node_modules
npm install
```
It is very heavy though, I really hope this gets fixed any time soon.

**Comment by NN---** [comment]
npm 5.3.0 (Node 8.4) doesn't show outdated packages

**Comment by thewashingtonhua** [comment]
Any ideas about what's causing this? it's been months since the release of node 8 & npm 5. and core features like install / uninstall / outdated still not working.

**Comment by NN---** [comment]
I found that if your package.json has exact version e.g. "gulp": "1.2.3" , npm outdated doesn't show whether you have a newer version.
Only if you put at least "^1.2.3" , then npm shows if a new version is available.

**Comment by ghost** [comment]
@NN but that is logical, if you have exact version how it can be outdated? That's why you need to set version range which checks if your package is outdated by checking it in your `node_modules`.

**Comment by jokeyrhyme** [comment]
@bsvipas but that is not the earlier behaviour: it always compared your current version with what was available in the registry (or registries)

**Comment by ghost** [comment]
@jokeyrhyme so it will not update it but it will be warning/info to you?

**Comment by jokeyrhyme** [comment]
@bsvipas yep, information only, `outdated` never actually updated or installed anything: https://docs.npmjs.com/cli/outdated

**Comment by ghost** [comment]
@jokeyrhyme I misread then, thanks for info!

**Comment by NN---** [comment]
@bsvipas I want to use exact version and manually update when I want.
Is there parameter to outdated command which always checks version from npm repository even if I have exact version in package.json?

Now I have to replace exact version with at least version , update and set exact version once again.

**Comment by jokeyrhyme** [comment]
@NN--- this issue was originally about `--global`, but it's likely that there's an underlying issue that effects both local and global uses of `npm update`, just a temporary regression

It's not a long-term solution, but try downgrading to 4.x: `npm install --global npm@4`
By installing a few different versions, you ought to be able to narrow down which versions behave the way you want, and which versions don't

**Comment by NN---** [comment]
AFAIR 4 worked the way I want.

**Comment by darkmorpher** [comment]
Status request @iarna @zkat

**Comment by itacode** [comment]
`npm outdated` outputs nothing.
I am on latest windows 10 with node 9.2.0 and npm 5.5.1.

**Comment by alienpavlov** [comment]
Still having issue with `npm outdated --global` it outputs nothing after 5 second of hanging.
macOS 10.13.3 and npm 5.6.0

**Comment by itacode** [comment]
`npm outdated` now works as expected.
Windows 10 node 9.5 npm 5.6

**Comment by akshaybabloo** [comment]
Not working in npm 5.7.1 as well

**Comment by MartinX3** [comment]
Using npm 5.8.0
Not every update able package is shown.

**Comment by alexdevero** [comment]
Two issues that are related:
1) `npm -g outdated` doesn't work properly - viz. 2)
2) `npm -g ls --depth=0` doesn't show globally installed packages but content of `nodejs` dir in `Program files`.
```
$ npm ls -g --depth=0
C:\Program Files\nodejs
`-- npm@5.8.0
```
npm: 5.8.0
nodejs: 9.11.1
OS: Windows 10 (1709)

**Comment by darkmorpher** [comment]
Still displaying persistent update notification even though this bug persists... :unamused:

![](https://i.imgur.com/c7Yotev.png)
